design/metric-math.md

```diff
@@ -0,0 +1,202 @@
+```
+m1 * 100          :: TS
+AVG(m1)           :: S
+AVG([m1, m2])     :: TS
+AVG([m1, m2]) + 1 :: TS
+SUM([m1, m2])     :: TS
+METRICS()         :: TS[]
+```
+
+Both operators and functions are polymorphic with respect to input and output types - the result type depends on the input type, like an overloaded function. This is important, because if we aim to reproduce a representation of this mathematical system in code, then we should also capture and enforce those rules as best we can.
+
+For example, the `AVG` function returns a `S` if passed a single `TS`, while it returns a `TS` if passed a `TS[]`:
+
+```
+AVG(m1)       :: S // average of all data points 
+AVG([m1, m2]) :: TS // average of the series' points at each interval
+```
+
+A consequence of this is you can not alarm on `AVG(TS)` because its type is `S`.
+
+## Intermediate Computation
+Metric definitions and math expressions co-exist in a list of `MetricDataQuery` objects. Only one time-series (single line) result may be compared against the alarm condition. That is to say, the result type of a query must be `TS` to be valid as an Alarm.
+
+This is achieved by setting the `ReturnData` flag. When `true`, the result of that expression - whether it be a `TS` or `TS[]` - materializes as lines on a graph; setting it to `false` means it doesn't. This mechanic then enables two applications: 
+* Reduction of a complex expression into smaller, simpler (and potentially re-usable) parts;
+* Identify the `TS` result you want to monitor in the alarm.
+
+Let's look at an example CloudFormation YAML definition of an alarm using metric-math:
+
+```yaml
+MyAlarm:
+  Type: AWS::CloudWatch::Alarm
+  Properties:
+    Metrics:
+      - Id: errors
+        MetricStat:
+          Metric:
+            MetricName: Errors
+            Namespace: Test
+          Period: 300
+          Stat: Sum
+        ReturnData: false
+      - Id: requests
+        MetricStat:
+          Metric:
+            MetricName: Requests
+            Namespace: Test
+          Period: 300
+          Stat: Sum
+        ReturnData: false
+      - Id: error_rate
+        Expression: errors / requests * 100
+        ReturnData: true
+```
+
+We have requested two specific metrics, `errors` and `requests`. This brings those metrics into the scope of the query with their respective IDs. Think of it like assigning local variables for use later on. Also note how the value of `ReturnData` is `false` for these metrics - as previously mentioned, only one entry in an alarm definition is permitted to return data. By specifying `false`, we are indicating that this value is an 'intermediate' value used only as part of the larger computation. The expression, `error_rate`, then computes the error rate as a percentage using a simple expression, `errors / requests * 100`. `ReturnData` is set to true because we want to alarm on this `TS` value.
+
```

Comment:
Love the overview section. It gave me enough context to be able to reason about the design without needing to go look up all the material!